docs/architecture/adr-028-public-key-addresses.md

```diff
@@ -0,0 +1,110 @@
+# ADR 028: Public Key Addresses
+
+## Changelog
+
+- 2020/08/18: Initial version
+
+## Status
+
+Proposed
+
+## Context
+
+Issue [\#3685](https://github.com/cosmos/cosmos-sdk/issues/3685) identified that public key
+addresses are currently overlapping. One initial proposal was extending the address length and
+adding prefixes for different types of addresses. 
+
+@ethanfrey explained an alternate approach originally used in https://github.com/iov-one/weave:
+
+> I spent quite a bit of time thinking about this issue while building weave... The other cosmos Sdk.
+
+> Basically I define a condition to be a type and format as human readable string with some binary data appended. This condition is hashed into an Address (again at 20 bytes). The use of this prefix makes it impossible to find a preimage for a given address with a different condition (eg ed25519 vs secp256k1).
+
+> This is explained in depth here https://weave.readthedocs.io/en/latest/design/permissions.html
+
+> And the code is here, look mainly at the top where we process conditions. https://github.com/iov-one/weave/blob/master/conditions.go
+
+And explained how this approach should be sufficiently collision resistant:
+> Yeah, AFAIK, 20 bytes should be collision resistance when the preimages are unique and not malleable. A space of 2^160 would expect some collision to be likely around 2^80 elements (birthday paradox). And if you want to find a collision for some existing element in the database, it is still 2^160. 2^80 only is if all these elements are written to state.
+
+> The good example you brought up was eg. a public key bytes being a valid public key on two algorithms supported by the codec. Meaning if either was broken, you would break accounts even if they were secured with the safer variant. This is only as the issue when no differentiating type info is present in the preimage (before hashing into an address).
+
+> I would like to hear an argument if the 20 bytes space is an actual issue for security, as I would be happy to increase my address sizes in weave. I just figured cosmos and ethereum and bitcoin all use 20 bytes, it should be good enough. And the arguments above which made me feel it was secure. But I have not done a deeper analysis.
+
+In discussions in [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694), we agreed to go with an
+approach similar to this where essentially we take the first 20 bytes of the `sha256` hash of
+the key type concatenated with the key bytes, summarized as `Sha256(KeyTypePrefix || Keybytes)[:20]`.
+
+## Decision
+
+### Legacy Public Key Addresses Don't Change
+
+`secp256k1` and multisig public keys are currently in use in existing Cosmos SDK zones. We
+don't want to change existing addresses. So the addresses for these two key types will remain the same.
```

Comment:
as cory pointed out, I think it would make this ADR clearer to state what these 2 key types use now:
- secp256k1: `ripemd160(sha256(pk_bytes))`
- legacy amino multisig: `sha256(aminoCdc.Marshal(pk))[:20]`